keps/sig-api-machinery/3610-env-injection/README.md

```diff
@@ -0,0 +1,920 @@
+For namespace-wide, it would be a problem to make an env to all namespaces. There should be a patched annotation after new namespace is created. 
+
+For cluster-wide: all pods in the cluster will get the envs according to the configmap or env of `kube-system` namespace.
+
+- for cluster-wide, name of the annotaiton or configmap can be `cluster-wide-envs`.
+
+**overwrite-or-not option support VS just overwrite**
+
+I defined `forceEnv` and `defaultEnv` in the API Object, so that we can choose to overwrite or not.
+If we want to just overwrite it, we may use just one like `env` or `defaultEnv`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A global setting like proxy or env that used by default in some clients.
+
+- This would also be very useful when running a service mesh like Linkerd within the cluster, as you have to set the http_proxy environment for everything that should use it.
+
+
+#### Story 2
+
+From <https://github.com/kubernetes/kubernetes/issues/48180#issuecomment-331941326>
+> Automatically inject the kubernetes namespace into an environment variable for me. With dozens of different kube templates people sometimes
+> forget to expose it manually (<https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/>)
+> Have compute nodes have special knowledge about themselves that can be injected into environment variables. We have an odd/even compute node
+> concept for upgrading openshift and this information may come in handy
+> Be able to specify system-wide environment variables. We have multiple openshift instances (different environments and different data centers).
+> It would be great if that data could be made automatically available without having to deal with custom env sections in configs.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+The design cannot fit the cluster-wide case.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+For initial design with annotations, as there is no validation for the key value in annotation or configmap, users will get errors util they create a new pod.
+For API Object, we can validate the key value in the API Object.
```

Comment:
- This may have a risk to overwrite system env like `KUBERNETES_PORT`. Should there be a disallowed list?

Reply:
If we allow env such as `KUBERNETES_SERVICE_HOST` to be overwritten, then should we not do the same for `KUBERNETES_PORT` also? But +1 for having a disallow list.

Reply:
https://github.com/kubernetes/kubernetes/blob/3d67e162a03d0d724dc5a15a0617c5e8572c7b4a/cmd/kubelet/app/options/options.go#L488

AllowedUnsafeSysctls may be something that we can follow. We can mark "KUBERNETES_*" as disallowed.

Reply:
So what would be the behavior if this is set ?

1. Say the new argument that disallows certain unsafe env is set to `KUBERNETES_*`
2. Someone tries to create a `GlobalEnv` resource with `KUBERNETS_SOMETHING: somevalue` as the env injection.

Do we error out and indicate that you can't even create the `GlobalEnv` resource until you cleanup the unsafe env or we accept the `GlobalEnv` but ignore the unsafe one during injection ?